### DISCUSSION TITLE
Interrupt custom action

### DISCUSSION THREAD

**Discussion by d-v** [wixtoolset/issues#6412] [Questions] [answered]
I've got a custom bootstrapper executing an msi package with several custom actions that take around 5 minutes. It's currently possible to cancel in between the custom actions and pressing cancel does usually kick in after a while but sometimes it seems to get stuck. Is there any way to signal from the bootstrapper to the install engine that I want to cancel the execution of the currently running custom action?

**Comment by robmen** [comment, marked as answer]
Yes. The custom actions need to send progress so they can be interrupted. Note, this is how custom actions can be canceled even if the package wasn't included in a bundle.

**Reply by d-v** [reply]
Got it. What would be a good way to send this progress? The two commands are just calls to wsl that happen to take several minutes each. I am not sure how to go about supporting aborting these calls when the user presses cancel. Or if the only option is to wait for the actions to complete and cancelling the execute sequence then.

**Reply by robmen** [reply]
Custom actions are specific to the action they are executing. Thus the name "custom action". You need to answer the questions you raised to yourself and design the custom actions to handle all those scenarios.

Custom actions are not easy to write.

**Reply by d-v** [reply]
```
<SetProperty Id="InstallDebian" Value="&quot;[#DebianExe]&quot; install --root"
            Before="InstallDebian" Sequence="execute"/>
<CustomAction Id="InstallDebian" BinaryKey="WixCA" DllEntry="WixQuietExec64"
	    Execute="deferred" Return="ignore" Impersonate="yes"/>
```

Here is an example of a custom action that is running for several minutes during which the user presses cancel. Is it possible to abort the custom action before it completes?

**Reply by robmen** [reply]
The code for the `WixQuietExec` custom action exists here: https://github.com/wixtoolset/wix3/blob/develop/src/ext/ca/wixca/dll/qtexecca.cpp

**Reply by d-v** [reply]
Thank you, I appreciate the help!

**Reply by nickhoeferpickpro** [reply]
can WixCA be used in wix v4?
When I try to use it, i get this error.
`error WIX0094: The identifier 'Binary:WixCA' could not be found. Ensure you have typed the reference correctly and t
hat all the necessary inputs are provided to the linker.`
I've read that it needs WixUtilExtension.dll.  Can I reference that with nuget?
I found this on nuget registry 
`<PackageReference Include="WixToolset.Util.wixext" Version="4.0.2" />`
but the nupkg actually has "WixToolset.Util.wixext.dll". 
Are they the same thing?

**Reply by nickhoeferpickpro** [reply]
Never mind, it looks like I just needed to use Wix4UtilCA_$(sys.BUILDARCHSHORT) instead

**Reply by sohamssys** [reply]
I am having a similar issue. I wanted to know how exactly do I send progress from custom actions? Where do I send it? I am fairly new to Wix and Custom actions. Thanks.
